elide-core/src/main/java/com/yahoo/elide/core/pagination/Pagination.java

```diff
@@ -23,77 +22,119 @@
+
+
+    private Pagination(Map<PaginationKey, Integer> pageData) {
+        this.pageData = pageData;
     }
 
     /**
      * Given json-api paging params, generate page and pageSize values from query params.
+     *
      * @param queryParams The page queryParams (ImmuatableMultiValueMap).
      * @return The new Page object.
      */
     public static Pagination parseQueryParams(final MultivaluedMap<String, String> queryParams)
             throws InvalidValueException {
         final Map<PaginationKey, Integer> pageData = new HashMap<>();
-        final MutableBoolean pageTotalsRequested = new MutableBoolean(false);
-
         queryParams.entrySet()
                 .forEach(paramEntry -> {
                     final String queryParamKey = paramEntry.getKey();
                     if (PAGE_KEYS.containsKey(queryParamKey)) {
-                        final String value = paramEntry.getValue().get(0);
-                        try {
-                            pageData.put(PAGE_KEYS.get(queryParamKey), Integer.parseInt(value, 10));
-                        } catch (NumberFormatException e) {
-                            throw new InvalidValueException("page values must be integers");
+                        PaginationKey paginationKey = PAGE_KEYS.get(queryParamKey);
+                        if (queryParamKey.equals(PAGE_TOTALS_KEY)) {
+                            // page[totals] is a valueless parameter, use value of 0 just so that its presence can
+                            // be recorded in the map
+                            pageData.put(paginationKey, 0);
+                        } else {
+                            final String value = paramEntry.getValue().get(0);
+                            try {
+                                int intValue = Integer.parseInt(value, 10);
+                                pageData.put(paginationKey, intValue);
+                            } catch (NumberFormatException e) {
+                                throw new InvalidValueException("page values must be integers");
+                            }
                         }
                     } else if (queryParamKey.startsWith("page[")) {
                         throw new InvalidValueException("Invalid Pagination Parameter. Accepted values are "
-                                + PAGE_KEYS_CSV
-                        );
+                                + PAGE_KEYS_CSV);
                     }
                 });
         return new Pagination(pageData).evaluate(null);
     }
 
-        if (pageData.isEmpty()) {
-            return DEFAULT_PAGINATION;
-        } else if (pageData.size() > 2) {
-            throw new InvalidValueException("Invalid usage of pagination parameters.");
-        }
+    /**
+     * Evaluates the pagination variables. Uses the Paginate annotation if it has been set for the entity to be
+     * queried.
+     *
+     * @param entityClass
+     */
+    public Pagination evaluate(final Class entityClass) {
```

Comment:
One thing I'm not sure about is that we will evaluate certain rules more than once.  Once parsing the query params, and then again when evaluate is called for a given entity class.  Maybe this could be split into two evaluation function.  One which validates limits (done per class), and one which evaluates the legal set of parameters - done at query param parsing time.